nixos/modules/services/security/bitwarden_rs/default.nix

```diff
@@ -0,0 +1,177 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.bitwarden_rs;
+  description = "Bitwarden Server (Rust Edition)";
+
+  nameToKey = name:
+    let
+      parts = builtins.split "([A-Z0-9]+)" name;
+      partsToKey = parts: removePrefix "_" (foldl' (key: x:
+        if isList x then key + "_" + head x
+        else if key != "" && elem (substring 0 1 x) lowerChars then # to handle e.g. [ "disable" [ "2FAR" ] "emember" ]
+          let last = stringLength key - 1;
+          in substring 0 last key + optionalString (substring (last - 1) 1 key != "_") "_" + substring last 1 key + toUpper x
+        else key + toUpper x) "" parts);
+    in partsToKey parts;
+
+  configFile = pkgs.writeText "bitwarden_rs.env" (concatMapStrings (s: s + "\n") (
+    (concatLists (mapAttrsToList (name: value:
+      if value != null then [ "${nameToKey name}=${if isBool value then boolToString value else toString value}" ] else []
+    ) (cfg.config // {
+      inherit (cfg) dataFolder;
+      webVaultEnabled = cfg.enableWebVault;
+      webSocketEnabled = cfg.enableWebsocket;
+    })))));
+
+in {
+  options.services.bitwarden_rs = with types; {
+    enable = mkEnableOption description;
+
+    user = mkOption {
+      type = str;
+      default = "bitwarden_rs";
+      description = ''
+        User under which ${description} runs.
+      '';
+    };
+
+    group = mkOption {
+      type = str;
+      default = "bitwarden_rs";
+      description = ''
+        Group under which ${description} runs.
+      '';
+    };
+
+    dataFolder = mkOption {
+      type = str;
+      default = "/var/lib/bitwarden_rs";
```

Comment:
I would also ask you to question if the flexibility of a `dataDir` is really required or not. Having this option really works against systemd and adds complication for something the user can take care of with symlinks if *really* necessary.

See https://github.com/NixOS/nixpkgs/pull/60019#discussion_r277558120 for more discussion.

Reply:
Actually, this option is part of bitwarden_rs's configuration, which I happen to reuse for the service. What trick is there again to work around needing a value from an attribute set to define a default of that attribute set, since that is what made me choose to put them outside the config option.

This produces infinite recursion (I get why, just do not remember how to work around it properly):
```nix
{
  services.bitwarden_rs.config.webVaultFolder = mkIf cfg.config.webVaultEnabled (mkDefault "${pkgs.bitwarden_rs-vault}/share/bitwarden_rs/vault");
}
```

Right now I worked around it with an apply.
```nix
{
  apply = config: config // optionalAttrs (config.webVaultEnabled && !(config ? webVaultFolder)) {
    webVaultFolder = "${pkgs.bitwarden_rs-vault}/share/bitwarden_rs/vault";
  };
}
```

@aanderse So you would suggest to not allow the user to configure the `dataFolder` of bitwarden_rs but always set it to `/var/lib/bitwarden_rs` and have the service always use `StateDirectory = "bitwarden_rs"`?

Reply:
@msteen Yes, for a service like this I suggest hard coding the data directory as that seems to work for almost every other Linux distribution out there. If the user has a really exotic use case then symlinks should more than suffice.

Reply:
Thanks for the confirmation. I have added an assertion to make clear to users that this is now the case, so that changes to `config.dataFolder` are not silently ignored.